app/src/main/java/net/gsantner/markor/format/todotxt/TodoTxtTextActions.java

```diff
@@ -96,112 +94,78 @@ int getFormatActionsKey() {
         @Override
         public void onClick(View view) {
             view.performHapticFeedback(HapticFeedbackConstants.KEYBOARD_TAP);
-            final SttCommander sttcmd = SttCommander.get();
             final String origText = _hlEditor.getText().toString();
-            final int origSelectionStart = _hlEditor.getSelectionStart();
-            final SttTaskWithParserInfo origTask = sttcmd.parseTask(origText, origSelectionStart);
             final CommonTextActions commonTextActions = new CommonTextActions(_activity, _hlEditor);
-
-            final Callback.a1<SttTaskWithParserInfo> cbUpdateOrigTask = (updatedTask) -> {
-                if (updatedTask != null) {
-                    SttCommander.SttTasksInTextRange rangeInfo = sttcmd.findTasksBetweenIndex(origText, origTask.getLineOffsetInText(), origTask.getLineOffsetInText());
-                    Editable editable = _hlEditor.getText();
-                    rangeInfo.startIndex = Math.max(rangeInfo.startIndex, 0);
-                    rangeInfo.endIndex = Math.max(rangeInfo.endIndex, 0);
-                    try {
-                        editable.delete(rangeInfo.startIndex, rangeInfo.endIndex);
-                    } catch (Exception e) {
-                        e.printStackTrace();
-                    }
-                    sttcmd.regenerateTaskLine(updatedTask);
-                    editable.insert(rangeInfo.startIndex, updatedTask.getTaskLine() + "\n");
-
-                    // Try to figure out new cursor pos
-                    int cursor = rangeInfo.startIndex + origTask.getCursorOffsetInLine();
-                    if (cursor != 0) {
-                        cursor += _hlEditor.getText().length() - origText.length(); // difference
-                    }
-                    if (cursor == _hlEditor.getText().length()) {
-                        cursor--; // Move to last char in text;
-                    }
-                    if (cursor >= 0 && cursor <= _hlEditor.getText().length()) {
-                        _hlEditor.setSelection(cursor);
-                    }
-                }
-            };
-
+            final TodoTxtTask[] selTasks = TodoTxtTask.getSelectedTasks(_hlEditor);
 
             switch (_action) {
                 case R.string.tmaid_todotxt_toggle_done: {
-                    origTask.setDone(!origTask.isDone());
+                    String replaceDone = "x ";
                     if (AppSettings.get().isTodoAddCompletionDateEnabled()) {
-                        origTask.setCompletionDate(SttCommander.getToday());
+                        replaceDone += TodoTxtTask.getToday() + " ";
                     }
-                    cbUpdateOrigTask.callback(origTask);
+                    runRegexReplaceAction(
```

Comment:
@gsantner this is how we can implement #1043 here.

Note that this has an issue:

if I mark a task as done and then unmark it, the _creation date_ is removed as well.

i.e.

`(C) 2020-01-01 foo` -> mark done `x 2020-01-01 foo` -> unmark done -> `x 2020-01-01 foo`. 

This seems to be the behavior of master as well so I have left it as is here.